ARCs/arc-0076.md

```diff
@@ -0,0 +1,64 @@
+---
+arc: 76
+title: H999 Account
+description: SHA512-256 999 account
+author: Ludovit Scholtz (@scholtz)
+discussions-to: https://github.com/algorandfoundation/ARCs/issues/208
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-06-12
+---
+
+## Abstract
+This standard specifies a computation for seed bytes for H999 account. For general adoption it is better for people to remember passphrase than mnemonic. With this standard person can hash the passphrase and receive the seed bytes for X25529 algorand account.
+
+## Motivation
+By providing a clear and precise computation process, H999 account empowers individuals to effortlessly obtain their seed bytes for algorand account. In the realm of practicality and widespread adoption, the standard highlights the immense advantages of utilizing a passphrase rather than a mnemonic. Remembering a passphrase becomes the key to unlocking a world of possibilities. With this groundbreaking standard, individuals can take control of their X25529 Algorand account by simply hashing their passphrase and effortlessly receiving the corresponding seed bytes. It's time to embrace this new era of accessibility and security, empowering yourself to reach new heights in the world of H999 accounts. Let this standard serve as your guiding light, motivating community to embark on a journey of limitless possibilities and unparalleled success.
+
+This standard seek the synchronization between wallets which may provide password protected accounts.
+
+## Specification
+
+Hash of the seed bytes is calculated with algorithm:
+
+```
+const init = `${data}`;
+let hash = sha512.sha512_256.hex(init);
+for (let i = 1; i <= 997; i++) {
```

Comment:
PBKDF2 has input Password, and Salt, and Iteration count.. how do you expect two different wallet service come to the same  output for seed for the account? Should the salt be fixed by the standard? (does not feel much secure) Or user should remember also salt?

Reply:
That's no better than using `passphrase || username` as the input data, right? It just makes the rainbow table bigger. (Which is a good thing, of course, but hashing it down to a "salt" doesn't help.)

I'm confused by this entire proposal.  We think it's a good idea to have accounts protected only by how complicated the user makes their passphrase?  Why would we put that decision in the hands of a brand new, and therefore presumably naive, user?

Reply:
> That's no better than using passphrase || username as the input data, right? It just makes the rainbow table bigger. (Which is a good thing, of course, but hashing it down to a "salt" doesn't help.)

You could do that too I think. But you also want to add a salt that is unique to this ARC on top of that to prevent generic rainbow tables to apply in our case. 

> I'm confused by this entire proposal. We think it's a good idea to have accounts protected only by how complicated the user makes their passphrase? Why would we put that decision in the hands of a brand new, and therefore presumably naive, user?

I agree, there is a great danger by allowing the user choose a password. I made

Reply:
well rainbow table can help you only if you hash once.. right?

Thats why in original idea is that we hash the input 999 times.. Also I have not seen the rainbow table if you hash something multiple times, but i may be mistaken..

it is not about the username or email.. It is about any password to be hashed and created from that the seed for algorand account. I perfectly understand that this brute force is the danger there, however in the scenario where you would have input ```let init = `ARC-76-${email}-${password}-999-SHA512/256`;``` the prefix and suffix can be used as the salt, and this account can be reproduced in two different wallets just by knowing the correct email and correct password. 

If you want to suggest different seed generation, please provide code for it. I have provided code for H999 account.

Or if you want to change the text of the standard, please state exactly what you want to change.

For the statement "by going over all the possible passphrases" .. This must be communicated by the wallets not to provide weak passwords. I hope you are not suggesting that because someone can brute furce hash, the hash is not secure enough. Whole proof of work has been built on top of it.

"You need a different salt by person" .. this is exactly what i do not want. I want one password generate exactly same account in two different wallets.

Reply:
> Thats why in original idea is that we hash the input 999 times.. 

Everything depends on the initial entropy of the password.
But nowadays hashing 999 times is not considered secure enough in most cases for multiple reasons.
The proper way is to use PBKDF2 or argon2.

> Also I have not seen the rainbow table if you hash something multiple times, but i may be mistaken..

Rainbow tables can be done with any type of hashing. Maybe they are not available off-the-shelve for this exact 999-hashing option. But someone can just pay the amount of money necessary to create it and then sell it.
If there are enough money to steal, this would be worth it.

> For the statement "by going over all the possible passphrases" .. This must be communicated by the wallets not to provide weak passwords. I hope you are not suggesting that because someone can brute furce hash, the hash is not secure enough. Whole proof of work has been built on top of it.

Actually proof of work is exactly showing that hashes can be brute-forced: this is exactly what Bitcoin miners do: brute-force SHA256.

> "You need a different salt by person" .. this is exactly what i do not want. I want one password generate exactly same account in two different wallets.

Using the email is a good option in that case.

Reply:
Concretely, the main change in the standard is replacing hashing 999 times by PBKDF2 / argon2 with proper parameters.
Exact salt generation is to be discussed. My proposal was to derive the salt from the email and some fixed value specified in the standard (that all wallets would be using).